jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/NiceDecompositionBuilder.java

```diff
@@ -0,0 +1,291 @@
+
+    // map from decomposition nodes to the vertex sets
+    protected Map<Integer, Set<V>> decompositionMap;
+
+    // the root of the tree
+    protected int root;
+
+    // next integer for vertex generation
+    private int nextInteger;
+
+    /**
+     * Constructor for all methods used in the abstract method. This constructor instantiates the
+     * tree of the decomposition, the map from tree vertices to vertex sets and adds the root to the
+     * tree.
+     */
+    protected NiceDecompositionBuilder()
+    {
+        // creating objects
+        decomposition = new DefaultDirectedGraph<>(DefaultEdge.class);
+        decompositionMap = new HashMap<>();
+        nextInteger = 0;
+    }
+
+    /**
+     * Getter for the next integer, which is not yet used as a node in the decomposition. 
+     * Supplies the add vertex methods with new vertices.
+     * 
+     * @return unused integer node for the decomposition
+     */
+    private int getNextInteger()
+    {
+        return nextInteger++;
+    }
+    
+    /**
+     * Method for adding a root node. This method should be called at the start of every computation.
+     * It creates a single node with the bag containing {@code rootElement}.
+     * 
+     * @throws UnsupportedOperationException when calling this method twice
+     * @param rootElement the content of the bag of the root
+     * @return the root node
+     */
+    protected Integer addRootNode(V rootElement)
+    {
+        if(!decomposition.vertexSet().isEmpty())
+            throw new UnsupportedOperationException("A root can not be created twice");
+        
+        root = getNextInteger();
+        Set<V> bag = new HashSet<>();
+        bag.add(rootElement);
+        decompositionMap.put(root, bag);
+        decomposition.addVertex(root);
+        return root;
+    }
+
+    /**
+     * Method for adding a new join node.<br>
+     * {@code node} is copied two times jLeft and jRight. {@code node} afterwards becomes the root
+     * of the subtree. jRight becomes the child of {@code node} and has the children of {@code node}
+     * as children. jLeft becomes a leaf and the child of {@code node}. This can be used to make a
+     * join node retrospectively to branch of this node. jRight continues the path while jLeft adds
+     * another path.<br>
+     * The time complexity of this method is in $\mathcal{O}(|b(node)|)$.
+     * 
+     * @param node which nodes should become a join node
+     * @return the new children of the join node; first/left element has no children, second/right
+     *         element has the children of {@code node}
+     */
+    protected Pair<Integer, Integer> addJoinNode(Integer node)
+    {
+        Set<V> currentVertexBag = null;
+
+        // create first/left new child of node
+        int vertexChildLeft = getNextInteger();
+        decomposition.addVertex(vertexChildLeft);
+        currentVertexBag = new HashSet<V>(decompositionMap.get(node));
```

Comment:
Please, remove generic type parameters on the right-hand side of the assignments.